### DISCUSSION TITLE
How do I remove `null` as a valid variant?

### DISCUSSION THREAD

**Discussion by VapidLinus** [joe-bell/cva#97] [Q&A] [answered]
I'm typing a button that has a variant called "variant" for accent/primary/secondary etc. I've made "accent" the default variant.

However, the type now includes a `null` variant:
![image](https://user-images.githubusercontent.com/4418470/214019862-aca9c137-303c-4c01-afd0-2ebc689eedaa.png)

If I leave out the "variant" prop, the styling defaults to the "accent" variant which is what I expect. But if I enter `null`, no styling is applied for the variant.

I do not want `null` to be an option. I only expect `undefined` along with the other variants to be an option. I do not want to make the property required.

The `null` variant also shows up in Storybook which is not what I want:
![image](https://user-images.githubusercontent.com/4418470/214020783-8b487c18-d202-45ad-9e02-778bec987411.png)

Is there a way to get rid of the `null` variant?

**Comment by mnlfischer** [comment]
Thats the same problem we have. 
``` typescript
declare type ConfigVariants<T extends ConfigSchema> = {
    [Variant in keyof T]?: StringToBoolean<keyof T[Variant]> | null;
};
```
@joe-bell why there is a null type defined to its prop. When creating types or interfaces, this is not an implicit behavior setting props to null unless you define it in my understanding.

Is this intended? This makes it really hard to use props defined by cva for dynamic props that needs to be set for underlying components.

**Reply by joe-bell** [reply]
Answered: https://github.com/joe-bell/cva/discussions/97#discussioncomment-4780872

**Comment by joe-bell** [comment, marked as answer]
**Update:** [currently exploring `"unset"` keyword in `cva@1.0.0`](https://github.com/joe-bell/cva/tree/main/packages/cva/src/index.ts)

---

The behaviour of `null` was defined to reflect the behaviour in Stitches.js — to disable the variant altogether — however I agree with you this is fairly confusing without that prior context 

Changing this would be a breaking change so I’ll need to figure out the best way to approach this and batch in with some other ideas I have

I’m thinking instead something like:

```ts
button({ variant: “unset” })
``` 

(which would be opt-in of course)

**Comment by JameelKhan9** [comment]
Having a similar issue as I am trying to use size as an index which null breaks, instead having to override the null type definition provided by CVA.

The current workaround is to use this typescript utility to wrap the VariantProps from CVA:
https://bobbyhadz.com/blog/typescript-remove-null-and-undefined-from-type#using-the-nonnullable-type-recursively

Edit: @joe-bell The current workaround for the null type causes issues with React Prop defaults, as typescript is no longer seeing the default value applied as optional.

**Reply by joe-bell** [reply]
Answered: https://github.com/joe-bell/cva/discussions/97#discussioncomment-4780872

**Comment by akomm** [comment]
I don't want `any` variants to be nullable and `some` to be required.

This is how I currently reduce the manual labor of doing it at once, without having to nest the types at place of usage:

```ts
import {VariantProps as CvaVariantProps} from "class-variance-authority"

type Require<T, K extends keyof T> = Omit<T, K> & Required<Pick<T, K>>
type NonNullableProps<T> = {[P in keyof T]: NonNullable<T[P]>}

// 1. just make them all nonNullable
export type VariantProps<F extends (...args: any[]) => any> = NonNullableProps<
  CvaVariantProps<F>
>

// 2. also speficy the required props
export type RequiredVariantProps<
  F extends (...args: any[]) => any,
  K extends keyof CvaVariantProps<F>
> = NonNullableProps<Require<CvaVariantProps<F>, K>>
```

Not sure how efficient it is and whether TS optimizes calculations, when it has the same expression twice (`CvaVariantProps<F>` in 2.)
But I don't see another way to do it in order for K to have validation and auto completion. 

usage:
```ts
// A all: non-nullable, "bar" and "foobar" required
type A = RequiredVariantProps<typeof foo, "bar" | "foobar">
// B all: non-nullable, none: required
type B = VariantProps<typeof foo> 
```
